src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITModule.cpp

```diff
@@ -158,54 +171,62 @@ void LLVMJITModule::finish(const std::string& main_func) {
-      // where possible
-      pass_manager.add(llvm::createEarlyCSEPass(
-          /*enable_mem_ssa=*/true));  // Catch trivial redundancies
-
-      pass_manager.add(llvm::createJumpThreadingPass());  // Thread jumps.
-      pass_manager.add(llvm::createCFGSimplificationPass());
-
-      // remove load/stores in PHIs if instructions can be accessed directly post thread
-      // jumps
-      pass_manager.add(llvm::createNewGVNPass());
-
-      pass_manager.add(llvm::createDeadStoreEliminationPass());
-      pass_manager.add(llvm::createLICMPass());
-
-      pass_manager.add(llvm::createInstructionCombiningPass());
-
-      // module passes
-      pass_manager.add(llvm::createPromoteMemoryToRegisterPass());
-      pass_manager.add(llvm::createGlobalOptimizerPass());
-
-      pass_manager.add(llvm::createCFGSimplificationPass());  // cleanup after everything
-
-      pass_manager.run(*module);
-      break;
-    // TBD other optimize level to be added
-    case LLVMJITOptimizeLevel::DEBUG:
-      // DEBUG : default optimize level, will not do any optimization
-      break;
-    default:
-      LOG(FATAL) << "Invalid optimize level.";
-      break;
+void LLVMJITModule::optimizeIR() {
+  if (co_.optimize_ir) {
+    llvm::ModuleAnalysisManager module_analysis_mgr;
+    llvm::LoopAnalysisManager loop_analysis_mgr;
+    llvm::FunctionAnalysisManager function_analysis_mgr;
+    llvm::CGSCCAnalysisManager cgscc_analysis_mgr;
+
+    llvm::PassBuilder pass_builder;
+    pass_builder.registerModuleAnalyses(module_analysis_mgr);
+    pass_builder.registerFunctionAnalyses(function_analysis_mgr);
+    pass_builder.registerLoopAnalyses(loop_analysis_mgr);
+    pass_builder.registerCGSCCAnalyses(cgscc_analysis_mgr);
+    pass_builder.crossRegisterProxies(loop_analysis_mgr,
+                                      function_analysis_mgr,
+                                      cgscc_analysis_mgr,
+                                      module_analysis_mgr);
+
+    llvm::ModulePassManager module_pass_mgr;
+    llvm::FunctionPassManager function_pass_mgr;
+    llvm::LoopPassManager loop_pass_mgr;
+
+    module_pass_mgr.addPass(
+        llvm::AlwaysInlinerPass(false));  // Inline all functions labeled as always_inline
+
+    function_pass_mgr.addPass(
+        llvm::SROA());  // mem ssa drops unused load and store instructions, e.g. passing
+                        // variables directly where possible
+    function_pass_mgr.addPass(llvm::EarlyCSEPass(true));  // Catch trivial redundancies
+    function_pass_mgr.addPass(
+        llvm::SimplifyCFGPass());  // remove load/stores in PHIs if instructions can be
+                                   // accessed directly post thread jumps
+    function_pass_mgr.addPass(llvm::NewGVNPass());
+    function_pass_mgr.addPass(llvm::DSEPass());
+
+    function_pass_mgr.addPass(llvm::InstCombinePass());
+    function_pass_mgr.addPass(llvm::PromotePass());
+
+    loop_pass_mgr.addPass(llvm::LICMPass());
```

Comment:
Do we need to add any loop vectorizer related pass?

Reply:
Yes